File: lista3/P1 - Dijkstra?.py
# DESISTI DE FAZER EM GO...
n, m = map(int, input().split())

adjacencyMatrix  = {}
def updateMatrix(a, b, w):
  a_exist = adjacencyMatrix.get(a)
  b_exist = adjacencyMatrix.get(b)

  if a_exist and b_exist:
    a_and_b_know_each_other = a_exist.get(b)
    if a_and_b_know_each_other:
      w = min(w, adjacencyMatrix[a][b])
    adjacencyMatrix[a][b] = w
    adjacencyMatrix[b][a] = w

  else:
    if a_exist:
      adjacencyMatrix[a][b] = w
      adjacencyMatrix[b] = {a: w}

    elif b_exist:
      adjacencyMatrix[b][a] = w
      adjacencyMatrix[a] = {b: w}
    
    else:
      adjacencyMatrix[a] = {b: w}
      adjacencyMatrix[b] = {a: w}

predecessor      = {}
for i in range(m):
  a, b, w = map(int, input().split())
  updateMatrix(a, b, w)
  predecessor[a] = None
  predecessor[b] = None

h = []
distanceToRoot   = {}
def relax(u, v):
  distanceToRootFromU = distanceToRoot.get(u)
  if distanceToRootFromU == None:
    distanceToRootFromU = float('inf')
  distanceThroughU = distanceToRootFromU + adjacencyMatrix[u][v]

  distanceToRootFromV = distanceToRoot.get(v)
  if distanceToRootFromV == None:
    distanceToRootFromV = float('inf')

  if distanceToRootFromV > distanceThroughU:
    distanceToRoot[v] = distanceThroughU
    predecessor[v] = u
    heappush(h, (distanceThroughU, v))

used = {}
from heapq import *
def extractMin():
  return heappop(h)

predecessor[1] = 1
distanceToRoot[1] = 0
heappush(h, (0, 1))
i = 0
while i < n:
  u = extractMin()
  uid = u[1]
  if used.get(uid):
    continue
  for nid, w in adjacencyMatrix[uid].items():
    if not used.get(nid):
      relax(uid, nid)
  used[u] = True
  i += 1

output = str(n)
nod = predecessor[n]
while nod != 1 and nod != None:
  output = str(nod) + " " + output
  nod = predecessor[nod]
output = "1 " + output

if nod != None:
  print(output)
else:
  print("-1")

File: lista3/P1 - Dijkstra?.go
# DESISTI DE FAZER EM GO 
# ABAIXO O RASCUNHO...

package main
 
import (
  "fmt"
  "bufio"
  "os"
  "math"
)
 
// Took from here: https://www.codementor.io/tucnak/using-golang-for-competitive-programming-h8lhvxzt3
var reader *bufio.Reader = bufio.NewReaderSize(os.Stdin, 10000000)
var writer *bufio.Writer = bufio.NewWriterSize(os.Stdout,10000000)
func printf(f string, a ...interface{}) { fmt.Fprintf(writer, f, a...) }
func scanf(f string, a ...interface{}) { fmt.Fscanf(reader, f, a...) }

type node struct {
  id             int
  index          int
  distanceToRoot int
  predecessor    int
}

var nodes []*node
var adjacencyMatrix [100][100]int
func main() {
  // STDOUT MUST BE FLUSHED MANUALLY!!!
  defer writer.Flush()
  var n, m int
  scanf("%d %d\n", &n, &m)
  var x, y, z int
  for i := 0; i < m; i++ {
    scanf("%d %d %d\n", &x, &y, &z)
    updateMatrix(x, y, z)
  }
  nodes := make([]*node, n + 1)
  nodes[1] = &node{1, 1, 0, 0}
  for i := 2; i <= n; i++ {
    nodes[i] = &node{i, i, math.MaxInt64, -1}
  }
  for i := 1; i < n; i++ {
    fmt.Printf("%d\n%v", i, nodes)
    u := extractMin()
    for j := 1; j <= n; j++ {
      if nodes[j] != nil && adjacencyMatrix[u.id][j] != 0 {
        relax(u.id, j)
      }
    }
    nodes[u.id] = nil
  }
}

// TODO
func extractMin(n int) *node {
  return nil
}

func relax(uid, vid int) {
  u, v := nodes[uid], nodes[vid]
  distanceThroughU := u.distanceToRoot + adjacencyMatrix[u.id][v.id]
  if v.distanceToRoot > distanceThroughU {
    v.distanceToRoot = distanceThroughU
    v.predecessor = u.id
  }
}

func min(a, b int) int {
    if a <= b { return a }
    return b
}

func updateMatrix(x, y, z int) {
    if adjacencyMatrix[x][y] != 0 {
      z = min(z, adjacencyMatrix[x][y])
    }
    adjacencyMatrix[x][y] = z
    adjacencyMatrix[y][x] = z
}
